Fix/Make proc_macro span_close() and span_open() more accurate after set_span() calls

Keith-Cancel opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149052)*
<!-- homu-ignore:end -->

When `Group::set_span()` is called the span's for `Group::span_open()` and `Group::span_close() ` methods just become the whole span instead where it should be the span of the delimiters. This PR ensures they have at least a more usable value after a call to `Group::set_span()`.

Here is a kinda of motivating example of this issue I observed. If you compare the current behavior the outputs the diagnostics will not be on the opening and closing delimiters after calling set span. If you look at the outputs after using this PR they will match. Further, afterward `Span::eq()` then evaluates to true which nice to see.

```rust
#![feature(proc_macro_diagnostic)]
#![feature(proc_macro_span)]
extern crate proc_macro;

use proc_macro::TokenStream;
use proc_macro::TokenTree;

#[proc_macro_attribute]
pub fn foo(args: TokenStream, _: TokenStream) -> TokenStream {
    let mut i = args.into_iter();
    let mut grp = match i.next().expect("iterator empty") {
    TokenTree::Group(g) => g,
        _ => panic!("Expected Group")
    };

    let o_open = grp.span_open();
    let o_close = grp.span_close();
    o_open.note("Old Open Span").emit();
    o_close.note("Old Close Span").emit();

    // Set the span with the same span
    grp.set_span(grp.span());

    // Notice where the span these diagnostics are compared to the first.
    let n_open = grp.span_open();
    let n_close = grp.span_close();
    n_open.note("New Open Span").emit();
    n_close.note("New Close Span").emit();


    grp.span().note(&format!(
        "old_open == new_open: {}\nold_close == new_close: {}",
        o_open.eq(&n_open),
        o_close.eq(&n_close)
    )).emit();

    return TokenStream::new();
}
```

This is all I did to invoke the proc_macro example above.
```rust
#[macro_test::foo({ bar })]
struct Foo;
```

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19458099213/job/55675964723) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/55675964723) [(plain)](https://github.com/rust-lang/rust/commit/5888bc1bafdfba3744914b1cfdf5e802038c26b7/checks/55675964723/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   --> src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/mod.rs:136:9

You can update all `expect!` tests by running:

    env UPDATE_EXPECT=1 cargo test

To update a single test, place the cursor on `expect` token and use `run` feature of rust-analyzer.

Expect:
----
SUBTREE $$ 42:Root[0000, 0]@0..100#ROOT2024 42:Root[0000, 0]@0..100#ROOT2024
  IDENT   ident 42:Root[0000, 0]@0..5#ROOT2024
  PUNCH   , [alone] 42:Root[0000, 0]@5..6#ROOT2024
  SUBTREE [] 42:Root[0000, 0]@7..8#ROOT2024 42:Root[0000, 0]@8..9#ROOT2024



SUBTREE $$ 42:Root[0000, 0]@0..100#ROOT2024 42:Root[0000, 0]@0..100#ROOT2024
  IDENT   ident 42:Root[0000, 0]@0..5#ROOT2024
  PUNCH   , [alone] 42:Root[0000, 0]@5..6#ROOT2024
  SUBTREE [] 42:Root[0000, 0]@7..9#ROOT2024 42:Root[0000, 0]@7..9#ROOT2024
----

Actual:
----
SUBTREE $$ 42:Root[0000, 0]@0..100#ROOT2024 42:Root[0000, 0]@0..100#ROOT2024
  IDENT   ident 42:Root[0000, 0]@0..5#ROOT2024
  PUNCH   , [alone] 42:Root[0000, 0]@5..6#ROOT2024
  SUBTREE [] 42:Root[0000, 0]@7..8#ROOT2024 42:Root[0000, 0]@8..9#ROOT2024



SUBTREE $$ 42:Root[0000, 0]@0..100#ROOT2024 42:Root[0000, 0]@0..100#ROOT2024
  IDENT   ident 42:Root[0000, 0]@0..5#ROOT2024
  PUNCH   , [alone] 42:Root[0000, 0]@5..6#ROOT2024
  SUBTREE [] 42:Root[0000, 0]@7..8#ROOT2024 42:Root[0000, 0]@8..9#ROOT2024
----

Diff:
----
SUBTREE $$ 42:Root[0000, 0]@0..100#ROOT2024 42:Root[0000, 0]@0..100#ROOT2024
  IDENT   ident 42:Root[0000, 0]@0..5#ROOT2024
  PUNCH   , [alone] 42:Root[0000, 0]@5..6#ROOT2024
  SUBTREE [] 42:Root[0000, 0]@7..8#ROOT2024 42:Root[0000, 0]@8..9#ROOT2024



SUBTREE $$ 42:Root[0000, 0]@0..100#ROOT2024 42:Root[0000, 0]@0..100#ROOT2024
  IDENT   ident 42:Root[0000, 0]@0..5#ROOT2024
  PUNCH   , [alone] 42:Root[0000, 0]@5..6#ROOT2024
  SUBTREE [] 42:Root[0000, 0]@7..98#ROOT2024 42:Root[0000, 0]@78..9#ROOT2024
----

---- tests::test_fn_like_macro_clone_ident_subtree stdout end ----

failures:
```

</details>

petrochenkov commented:
`DelimSpan::from_single` in the compiler also currently doesn't try to update the opening and closing spans in a smart way, just sets the whole span.

I think it would be better to test the opening/closing span recovery strategy on the compiler first, and then do the same for proc macros.
There are multiple alternatives, and in some of them the choices here are obvious, but in some not and need to be checked experimentally (what diagnostic changes they cause).
- The delimiter is invisible
  - It's not clear what is better to use, the whole span, or empty spans at the end and beginning, need to check experimentally
- The delimiter is visible and the source map actually contains the matching symbols at the beginning and end of the span
  - The obvious choice - use spans of those symbols
- The delimiter is visible, but the source map doesn't contain the matching symbols neither at the beginning nor at the end
  - It's not clear what is better to use, the whole span, or empty spans at the end and beginning, need to check
- The delimiter is visible, but the source map only contains one matching symbols either at the beginning or at the end, the weird case
  - It's still not clear what is better to use for both the matching delimiter (its symbol's span, whole span, or empty span) and the non-matching delimiter (whole span or empty span), need to check

petrochenkov commented:
>I suppose we could inspect parts of the stream itself if that is desired.

>... but there are no delimiters in the token stream

I was not suggesting to inspect token stream, "source map" is some source text to which all spans map, not tokens.

Keith-Cancel commented:
> Could you update the tests (to make CI green) and submit the compiler changes in a separate PR? @rustbot author

Here is the separate PR as you asked: https://github.com/rust-lang/rust/pull/149112

I would say a lot the differences on those tests that changed are caused by using `DelimSpan::open` or `Delimnspan::close` instead of `DelimSpand::entire()` somewhere when generating the diagnostic output. They also all seem to involve macro metavariables.

@rustbot ready

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/149112.
@rustbot blocked

Keith-Cancel committed:
refresh labels

Keith-Cancel committed:
Make `Group::span_open()` and Group::span_close()` more accurate after set_span call.

This tries to recover the open in and close delimiter spans for the new span, otherwise it reverts to old behavior of using the the whole span for the open and close delimiter.

Veykril commented on library/proc_macro/src/lib.rs:
```diff
@@ -880,7 +880,47 @@ impl Group {
+
+        // No source text to check or source text is too small.
+        // just go back to default behavior.
+        let source = match span.0.source_text() {
```
This seems like quite the expensive operation to call here. For one this goes through the bridge (which is expensive for rust-analyzer, especially this call, as once we implement it will always do an RPC roundtrip), but that aside, this also always allocates a new string.

petrochenkov replied:
>to at least to make sure the spans are from the same file

That also requires going through the bridge though.

SpriteOvO commented:
@Keith-Cancel Hi, ping from triage team. This PR has been inactive for a while. Are there any updates on this? Thanks.

Keith-Cancel commented:
> @Keith-Cancel Hi, ping from triage team. This PR has been inactive for a while. Are there any updates on this? Thanks.

Hello, @SpriteOvO !

I have been kinda swamped with other things so have not had a chance to get back to this.  Although I am kinda leaning on towards something akin to `set_span_open_close()`

I might just close this for now, if I get back to I can re-open it later.
